// adminjs-sql/lib/adapter/Property.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Property = void 0;
const adminjs_1 = require("adminjs");
class Property extends adminjs_1.BaseProperty {
    constructor(columnName, _isTitle, _isVisible, _isEditable, _availableValues, _isRequired, _isSet, _referenceTableName, _referenceColumnName, position, type, isId, isSortable) {
        super({
            path: columnName,
            isId,
            isSortable,
            position,
            type,
        });
        this.columnName = columnName;
        this._isTitle = _isTitle;
        this._isVisible = _isVisible;
        this._isEditable = _isEditable;
        this._availableValues = _availableValues;
        this._isRequired = _isRequired;
        this._isSet = _isSet;
        this._referenceTableName = _referenceTableName;
        this._referenceColumnName = _referenceColumnName;
    }
    name() {
        return this.columnName;
    }
    path() {
        return this.columnName;
    }
    isTitle() {
        return this._isTitle;
    }
    isVisible() {
        return this._isVisible;
    }
    isEditable() {
        return this._isEditable;
    }
    reference() {
        return this._referenceTableName;
    }
    availableValues() {
        return this._availableValues;
    }
    isArray() {
        return this._isSet;
    }
    isDraggable() {
        return this._isSet;
    }
    subProperties() {
        return [];
    }
    isRequired() {
        return this._isRequired;
    }
}
exports.Property = Property;
